docs/documentation.md

```diff
@@ -260,3 +260,36 @@ For a given weight within the node the update process is:
+
+* Allowing users to further understand automatic differentation via an extended graph mode
+* Implementation of the Hessian matrix to provide users an efficient method of calculating second order derivatives
+
+### Extended Understanding Through Graph Modes
+
+Within `pyad` already exists a basic graphing class which allows the user to delve into how the reverse mode splits their function into elementary operations and calculates the final value of the function. This is useful in the sense that it gives a visual intuition for what is going on behind the scenes in the reverse mode of AD. What would be more useful however is to expand this feature to also display a graph of the derivative information - showing the flow of the derivatives as they are calculated during the forward step of the reverse mode. This secondary graph would then give users a complete understanding of both the function value and derivative values through the evaluation trace.
+
+Also to further help with understanding it would be useful to improve the visuals to show which nodes are the original variables, coloured and labeled appropriately (x, y, z for instance) and also provide a companion trace table of the values to go alongside the graph.
+
+### Calculation of the Hessian
+
+Across science and engineering there are numerous applications which require knowledge of the second order derivatives of a function. For example in image processing there is a method known as Hessian feature detection. In Hessian feature detection the second order derivative of a window of pixels is calculated along the x and y coordinate spaces to provide an understanding of the change in pixel intensities. Thresholding these second order derivative values in comparison to their neighbours results in the detection of important image features such as corners and other areas of high texture. The determinant of the Hessian is also commonly used in optimization (such as in fluid engineering to solve a series of ordinary differential equations). 
+
+To calculate the Hessian matrix we would need to differentiate the function twice with respect to each input variable. For example taking a simple function:
+
+<img src=2nd_der_orig.png width="170">
+
+The first order derivatives would be:
+
+<img src=2nd_deriv_1st_deriv.png width="270">
+
+And then the second order derivatives would be:
+
+<img src=2nd_deriv_2nd_deriv.png width="500">
+
+
+To implement this second order derivative we could use either a double forward pass system which used the forward mode twice to calculate the second order derivatives (or similarly double reverse pass) or use a combination method such as forward mode then reverse mode. Using forward mode twice has a scales in complexity of `O(n^2)` as each pass of forward mode scales linearly with the number of variables `n`. Performing the forward then reverse method would scale in complexity of `O(n*m)` as the reverse mode scales linearly with the number of output variables `m`.
```

Comment:
I think both forward mode and reverse mode would have `O(n^2)` complexity for computing the hessian of a scalar output variable.

The `m` in the reverse mode here would just be `n` since the output variables are the `n` partials w.r.t. each input.